File: services/tradelog/tradelog-api/src/main/java/com/example/tradelog/api/service/DividendJournalService.java
package com.example.tradelog.api.service;

import com.example.tradelog.api.converter.TradeSummaryListConverter;
import com.example.tradelog.api.repository.DividendRepository;
import com.example.tradelog.api.spec.model.DividendJournalModel;
import com.example.tradelog.api.spec.model.TradeSummaryModel;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Map;
import java.util.Optional;

@Service
public class DividendJournalService {

    private static final Logger log = LoggerFactory.getLogger(DividendJournalService.class);

    private DividendRepository dividendRepository;

    public DividendJournalService(DividendRepository dividendRepository) {
        this.dividendRepository = dividendRepository;
    }

    public List<DividendJournalModel> getAllBySymbol(String accountId, String symbol) {
        return dividendRepository.getAllBySymbol(accountId, symbol);
    }

    public Map<String, TradeSummaryModel> getSummaries(String accountId) {
        List<TradeSummaryModel> modelList = dividendRepository.getSummaries(accountId);
        return TradeSummaryListConverter.toMap.apply(modelList);
    }

    public Optional<DividendJournalModel> createDividendRecord(String transactionId, DividendJournalModel model) {
        dividendRepository.createRecord(transactionId, model);
        return dividendRepository.getByTransactionId(transactionId);
    }
}

File: services/tradelog/tradelog-api/src/main/kotlin/com/example/tradelog/api/core/service/DividendJournalService.kt
package com.example.tradelog.api.core.service

import arrow.core.Either
import arrow.core.flatMap
import com.example.common.service.ServiceError
import com.example.common.service.toServiceError
import com.example.tradelog.api.core.converter.TradeSummaryUtil
import com.example.tradelog.api.core.model.DividendJournalModel
import com.example.tradelog.api.core.model.TradeSummaryModel
import com.example.tradelog.api.db.repository.DividendJournalRepository
import org.springframework.stereotype.Service
import java.util.*

@Service
class DividendJournalService(private val repository: DividendJournalRepository): JournalService<DividendJournalModel> {

    override fun getSummaries(accountId: UUID): Either<ServiceError, Map<String, TradeSummaryModel>> {
        return repository.getSummaries(accountId)
                .mapLeft(::toServiceError)
                .map { TradeSummaryUtil.toMap(it) }
    }

    override fun getAllBySymbol(accountId: UUID, portfolioId: UUID, symbol: String): Either<ServiceError, List<DividendJournalModel>> {
        return repository.getAllBySymbol(accountId, portfolioId, symbol)
                .mapLeft(::toServiceError)
    }

    override fun createRecord(transactionId: UUID, model: DividendJournalModel): Either<ServiceError, DividendJournalModel> {
        val createRecord = repository.createRecord(transactionId, model)
                .mapLeft(::toServiceError)

        val readRecord: (Unit) -> Either<ServiceError, DividendJournalModel> = {
            repository.getById(model.transactionDetails.accountId, transactionId)
                    .mapLeft(::toServiceError)
        }

        return createRecord
                .flatMap(readRecord)
    }

    override fun editRecord(model: DividendJournalModel): Either<ServiceError, Unit> {
        return repository.editRecord(model)
                .mapLeft(::toServiceError)
    }

    override fun deleteRecord(transactionId: UUID): Either<ServiceError, Unit> {
        return repository.deleteRecord(transactionId)
                .mapLeft(::toServiceError)
    }

    override fun getById(accountId: UUID, transactionId: UUID): Either<ServiceError, DividendJournalModel> {
        return repository.getById(accountId, transactionId)
                .mapLeft(::toServiceError)
    }
}
